### PR TITLE
Update the repository specification

### PR DESCRIPTION
**Pull request by jonasfj**
* Documented authentication with `dart token add <hosted-url>`
 * Documented publishing with `dart pub publish`
 * Clean-up language centering on `<hosted-url>` and "package repository"
 * Removed discussion of meta-data only specific to `pub.dev`
 * Mentioned `hosted` in `pubspec.yaml`
 * Added notes on temporary URL usage and leavering GCS, S3 or similar blob storage service.

/cc: @isoos, @themisir, please see if you agree with this characterization of the repository specification.

### CODE REVIEW

File: doc/repository-spec-v2.md [hunk 1]
```diff
@@ -1,39 +1,216 @@
...
+Hence, a server shall not send `401` in case where a token is valid, but does
+not have permissions to access the package in question.
+
+When receiving a `401` response the `dart pub` client shall:
+ * Abort the current operation (exiting non-zero),
+ * Delete any token it may have stored for the given `<hosted-url>`,
+ * Inform the user that authentication is required,
+ * Print the `<message>` provided by the server.
+
+The `<message>` allows a custom _package server_ to how a token may be obtained.
```
Comment (themisir) [inline doc/repository-spec-v2.md:159]: The sentence seems incomplete. I might be wrong, but I think it needs ".. to inform user how a token..."

File: doc/repository-spec-v2.md [hunk 2]
```diff
@@ -68,16 +245,144 @@ Adding a custom `User-Agent` that allows _pub repository operators_ to identify
 }
 ```
 
-To fetch the package archive an HTTP `GET` request _following retries_ must be
+To fetch the package archive an HTTP `GET` request _following redirects must be
```
Comment (themisir) [inline doc/repository-spec-v2.md:248]: Markdown italic markup (`_`) is not terminated

**Review by themisir — COMMENTED** [review]
Also afaik json values doesn't allow trailing comma unlike dart or javascript but this doc contains multiple json snippets with trailing comma after last entey in map or array.

**Review by isoos — APPROVED** [review]
lgtm.
Do you have a short list of things in the spec that pub.dev may not do today? The `WWW-Authenticate` header is one them IIRC.

File: doc/repository-spec-v2.md [hunk 3]
```diff
@@ -1,39 +1,216 @@
...
-with the environment variable `PUB_HOSTED_URL`. In the rest of this document we
-shall refer to the base URL as `PUB_HOSTED_URL`.
+A package repository is a server from which packages can be downloaded,
+the default package repository is `'https://pub.dartlang.org'`, with public
+interface hosted at [pub.dev](https://pub.dev).
+
+## Hosted URL
+A custom package repository is identified by a _hosted-url_, like
+`https://pub.dartlang.org` or `https://some-server.com/prefix/pub/`.
+The _hosted-url_ always includes protocol `http://` or `https://`.
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply doc/repository-spec-v2.md:13]: That's correct, but we still allow `http`, we just don't allow authentication.

[the remark this answers is not in the document]
Comment (jonasfj) [reply doc/repository-spec-v2.md:13]: I'll add a note about this...

**Review by isoos — APPROVED** [review]

File: doc/repository-spec-v2.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (themisir) [reply doc/repository-spec-v2.md:13]: > That's correct, but we still allow `http`, we just don't allow authentication.

Oh true - just thought this spec is just about authentication for a moment 😅

**Pull request merged**